### ISSUE TITLE
The "vi" applet creates files with Windows style EOLs

### ISSUE THREAD

**Issue by ildar-shaimordanov** [rmyorston/busybox-w32#220]
The "vi" applet creates files with WIndows style EOLs. The same behavior is observed on both BusyBox and BusyBox64.

Windows_NT 10.0
BusyBox v1.34.0-FRP-3902-g61e53aa93

**Comment by rmyorston** [comment]
Yes, that's intentional. It's been that way since commit e6bb3e5823.

**Comment by ildar-shaimordanov** [comment]
So it seems will never be fixed. Can you explain the reason why it was developed in this way? Thank you.

**Comment by rmyorston** [comment]
I've looked back through my logs but couldn't find any explanation for the change. It must have seemed important at the time.

```vim``` has ```fileformat``` and ```fileformats``` options to control how line endings are handled. It's possible those would be useful on Windows. Upstream BusyBox is less likely to be concerned about such things.

**Comment by avih** [comment]
IMHO, if you want to automate it, I'd suggest to default to unix line ending, unless the opened file's first (if any) newline is preceded by `\r`.

**Comment by ildar-shaimordanov** [comment]
The following example shows, why I think it is a bug.

Here I create a file, check that it's unix style, edit it in busybox/v` and check that EOL are changed. Honestly, it's even enough to save the file without any changing in busybox/vi.

```
U:\>busybox sh -c "printf '%s\n' 1 2 > zzz"

U:\>busybox od -c zzz
0000000   1  \n   2  \n
0000004

U:\>busybox vi zzz

U:\>busybox od -c zzz
0000000   1  \r  \n   2  \r  \n   3  \r  \n
0000011
```

Other vi/vim implementations (for example gow/vim) work properly and save

**Comment by rmyorston** [comment]
It's not a bug; it's a design decision.

nano from Gow behaves like busybox-w32 vi. vim from Gow can be made to behave similarly by using non-default settings: ```vim -c 'set ff=dos ffs='```.

**Comment by avih** [comment]
> It's not a bug; it's a design decision.

Clearly it's a design decision. Whether or not it's a bug is in the eyes of the beholder, which also applies to the definition of "bug".

Either way, even if some applets have relaxed or altered behavior when taking CRLF files as input, the standard POSIX line ending is LF only, and personally I'd expect POSIX tools to produce POSIX compliant files, even on Windows.

Additionally, LF-only EOL is widely accepted on windows too, and I don't think any native Windows tool has issues with it currently, Notepad included.

As for `vi`, if you insist on supporting saved files with CRLF without explicit option, then IMHO, again, only do that if the input file already has CRLF.

**Comment by JuniorJPDJ** [comment]
> Either way, even if some applets have relaxed or altered behavior when taking CRLF files as input, the standard POSIX line ending is LF only, and personally I'd expect POSIX tools to produce POSIX compliant files, even on Windows.

Even if not, I'd just not change line endings in already existing file as it may break lots of things.
And afaik it's current behaviour.

**Comment by rmyorston** [comment]
I've implemented a cut-down version of the ```fileformats``` and ```fileformat``` options. The ```vi``` applet now checks the line endings of files opened for edit. If only LF line endings are detected it sets ```fileformat``` to ```unix```, otherwise ```dos```. Files are written with line endings appropriate to their ```fileformat```. Thus the format of Unix-style files is preserved.

```fileformats``` defaults to ```dos,unix``` but can be changed to ```unix,dos```. This only affects the type allocated when an empty edit buffer is created.

The ```fileformat``` option is read only.

The latest [prerelease](https://frippery.org/files/busybox/prerelease) has these changes.

**Comment by avih** [comment]
How can fileformats be changed to prefer unix when creating new files, if it's read-only?

**Comment by rmyorston** [comment]
>How can fileformats be changed to prefer unix when creating new files, if it's read-only?

```fileformats``` isn't read-only. ```fileformat``` is the one that's read-only.
